### PR TITLE
CB-16542 Improve logging in the role count validator

### PR DESCRIPTION
**Pull request by topolyai5**

### CODE REVIEW

**Review by foldik — APPROVED** [review]
Looks good.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java [hunk 1]
```diff
@@ -137,16 +148,25 @@ private void validateRequiredRoleCountInCluster(
...
                     .reduce(Integer::sum)
                     .ifPresent(remainingInstances -> {
+                        LOGGER.debug("{} remaining instances, required: {} in groups of {} for {}", remainingInstances, requiredCount, groupNames, role);
                         if (remainingInstances < requiredCount) {
                             throw new BadRequestException(String.format(
-                                    "Scaling adjustment is not allowed, based on the template it would eliminate all the instances with "
-                                            + "%s role which is not supported.", role));
+                                    "Scaling adjustment is not allowed in groups of %s, based on the template it would eliminate all the instances with "
+                                            + "%s role which is not supported. %s remained instances but %s the required", groupNames, remainingInstances,
+                                    requiredCount, role));
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java:169]: * `role` this should come right after `groupNames`.
* `%s remained instances but %s the required` -> `%s remaining instances but %s are required`

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java [hunk 2]
```diff
@@ -111,11 +112,20 @@ private void validateRequiredRoleCountInCluster(
...
+                List<InstanceGroup> filteredInstanceGroups = instanceGroups.stream()
                         .filter(instanceGroup -> instanceGroupsWithRoles.contains(instanceGroup.getGroupName()))
-                        .map(InstanceGroup::getNodeCount)
+                        .collect(Collectors.toList());
+                String groupNames = filteredInstanceGroups.stream().map(InstanceGroup::getGroupName).collect(Collectors.joining(","));
+                LOGGER.debug("Instance groups with role: {}", groupNames);
+                filteredInstanceGroups.stream()
+                        .map(instanceGroup -> {
+                            int nodeCount = instanceGroup.getNodeCount();
+                            LOGGER.debug("There are {} node(s) has in group of {} with role: {}", nodeCount, instanceGroup.getGroupName(), role);
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java:123]: Nit: delete `has`.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java [hunk 3]
```diff
@@ -137,16 +147,25 @@ private void validateRequiredRoleCountInCluster(
...
+            List<InstanceGroup> filteredInstanceGroups = instanceGroups.stream()
                     .filter(instanceGroup -> instanceGroupsWithRoles.contains(instanceGroup.getGroupName()))
-                    .map(instanceGroup ->
-                            instanceGroup.getNodeCount() + instanceGroupAdjustments.getOrDefault(instanceGroup.getGroupName(), 0))
+                    .collect(Collectors.toList());
+            String groupNames = filteredInstanceGroups.stream().map(InstanceGroup::getGroupName).collect(Collectors.joining(","));
+            LOGGER.debug("Instance groups with role: {}", groupNames);
+            filteredInstanceGroups.stream().map(instanceGroup -> {
+                        int nodeCount = instanceGroup.getNodeCount() + instanceGroupAdjustments.getOrDefault(instanceGroup.getGroupName(), 0);
+                        LOGGER.debug("There are {} node(s) has in group of {} with role: {}, adjusted node count: {}", instanceGroup.getNodeCount(),
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java:157]: Nit: delete `has`.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java [hunk 4]
```diff
@@ -137,16 +147,25 @@ private void validateRequiredRoleCountInCluster(
...
                     .reduce(Integer::sum)
                     .ifPresent(remainingInstances -> {
+                        LOGGER.debug("{} remaining instances, required: {} in groups of {} for {}", remainingInstances, requiredCount, groupNames, role);
                         if (remainingInstances < requiredCount) {
                             throw new BadRequestException(String.format(
-                                    "Scaling adjustment is not allowed, based on the template it would eliminate all the instances with "
-                                            + "%s role which is not supported.", role));
+                                    "Scaling adjustment is not allowed in groups of %s, based on the template it would eliminate all the instances with "
+                                            + "%s role which is not supported. %s remaining instances but %s are required", groupNames, remainingInstances,
+                                    role, requiredCount));
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateValidator.java:168]: * Argument `role` should come right after `groupNames`.
* `%s remaining instances but %s are required` both placeholders should be replaced with `%d`.

**Review by lajosrodek — CHANGES_REQUESTED** [review]

**Comment by topolyai5** [comment]
@foldik integrated into https://github.com/hortonworks/cloudbreak/pull/12473

**Pull request closed without merging**